user/src/com/google/gwt/user/server/rpc/logging/ReflectiveLoggerProvider.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright 2025 GWT Project Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.google.gwt.user.server.rpc.logging;
+
+import java.lang.reflect.Method;
+
+/**
+ * Uses reflection to create loggers that may be unavailable in some environments.
+ * Compatible with platform logging and Log4j; incompatible with SLF4j.
+ */
+abstract class ReflectiveLoggerProvider implements RpcLoggerProvider {
+
+  protected Method createLogger;
+  protected Object infoLevel;
+  protected Object warnLevel;
+  protected Object errorLevel;
+  protected Method logMessage;
+  protected Method logThrowable;
+  protected boolean available;
+
+  @Override
+  public RpcLoggerDelegate createLogger(String name) {
+    try {
+      Object logger = createLogger.invoke(null, name);
+      return new ReflectiveLogger(logger, infoLevel, warnLevel, errorLevel, logMessage,
+        logThrowable);
+    } catch (Exception e) {
+      throw new IllegalStateException(getClass() + " is not available on this platform", e);
+    }
+  }
+
+  @Override
+  public boolean isAvailable() {
+    return available;
+  }
+
+  protected Object getLogLevel(Class<?> levelClass, String levelName) {
+    for (Object level : levelClass.getEnumConstants()) {
+      if (level.toString().equals(levelName)) {
+        return level;
+      }
+    }
+    return null;
+  }
```

Comment:
Since we already insist this is an enum (rather than a class that has static/final fields), maybe just use Enum.valueOf(Class, String)? It won't matter at all for perf, but you could effectively inline it away that way. You'll need a slightly better cast, but arguably you need one anyway since we're asserting that the Class is an enum type.